# Data exploration

This section deals with the exploration of the dataset. First of all, the demographic data of the sample is described with regard to basic demographic variables. Second,the **macro- and microstructural** data is explored, comparing how the control and patient group might differ respectively. 

## 1. Demographic data

In the Dublin sample, there is a different subset of patients with **micro-structural (MD and FA)** data compared to those with **macro-structural (CT)** data. Both "subsets" are loaded and compared with regard to basic demographic variables.

### 1.1 Reading and adjusting the macro- and microstructural data

First, relevant modules are imported to read the data. Since different operating systems might differ in the way indicating file paths, the os.pardir() function is used to ensure that the code can run independently of various operating systems.

In [41]:
#import module to read data

import pandas as pd
import os

#store CT data in variable "CT_Dublin"

CT_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_thickness_Dublin.csv')
CT_Dublin = pd.read_csv(CT_Dublin_path)

To get the columns of our dataframe in order to see which variables are measured, we simply run the following command. 

In [42]:
CT_Dublin.columns

Index(['Subject ID', 'Age', 'Sex', 'Group', 'lh_bankssts_part1_thickness',
       'lh_bankssts_part2_thickness',
       'lh_caudalanteriorcingulate_part1_thickness',
       'lh_caudalmiddlefrontal_part1_thickness',
       'lh_caudalmiddlefrontal_part2_thickness',
       'lh_caudalmiddlefrontal_part3_thickness',
       ...
       'rh_supramarginal_part5_thickness', 'rh_supramarginal_part6_thickness',
       'rh_supramarginal_part7_thickness', 'rh_frontalpole_part1_thickness',
       'rh_temporalpole_part1_thickness',
       'rh_transversetemporal_part1_thickness', 'rh_insula_part1_thickness',
       'rh_insula_part2_thickness', 'rh_insula_part3_thickness',
       'rh_insula_part4_thickness'],
      dtype='object', length=312)

As we can see, besides the different brain regions there are columns that indicate variables containing the demographic data that we are interested in for now. To make things easier, we adjust the dataframe in selecting a subset that only carries the demographic variables.

In [43]:
#select a subset of the CT data with demographic variables

demographic_CT = CT_Dublin[["Subject ID", "Age", "Sex", "Group"]]

Having the subset of demographic information for the **macrostructural** data, the identical procedure is done for the **microstructural** data. So first, the data is read and adjusted to demogrpahic variables only.

In [44]:
#read MD and FA data

MD_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_MD_cortexAv_mean_Dublin.csv')
MD_Dublin = pd.read_csv(MD_Dublin_path)

FA_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_FA_cortexAv_mean_Dublin.csv')
FA_Dublin = pd.read_csv(FA_Dublin_path)

In [45]:
#select a subset of the MD and FA data with demographic variables

demographic_MD = MD_Dublin[["Subject ID", "Age", "Sex", "Group"]]
demographic_FA = FA_Dublin[["Subject ID", "Age", "Sex", "Group"]]

Normally, the shape of the dataframes for each MRI metric should be the same since it is only one sample. However, to check this we can run the following commands. 

In [46]:
demographic_CT.shape, demographic_MD.shape, demographic_FA.shape

((108, 4), (115, 4), (115, 4))

As we can see, the shapes are not identicial except for the **MD** and **FA** dataframe. Since the latter are both central characteristics of diffusion tensors (see further below), it makes sense that they exhibit the identical shape and were measured for the same amount of subjects. However, for the **CT** data, it seems like there are fewer rows indicating not the same amount of subjects for whom the metric was measured. So having a look at the dataframes should show that.

In [47]:
demographic_CT

Subject ID  Age  Sex  Group
0      CON9225   21    2      1
1      CON9229   28    2      1
2      CON9231   29    2      1
3     GASP3037   61    1      2
4     GASP3040   47    1      2
..         ...  ...  ...    ...
103    RPG9019   31    1      2
104    RPG9102   42    2      2
105    RPG9119   41    1      2
106    RPG9121   51    1      2
107    RPG9126   56    1      2

[108 rows x 4 columns]

In [48]:
demographic_MD

Subject ID  Age  Sex  Group
0      CON3140   37    2      1
1      CON3891   33    2      1
2      CON4664   40    2      1
3      CON7009   21    1      1
4      CON7024   59    1      1
..         ...  ...  ...    ...
110    RPG9102   42    2      2
111    RPG9103   37    1      2
112    RPG9119   41    1      2
113    RPG9121   51    1      2
114    RPG9126   56    1      2

[115 rows x 4 columns]

In [49]:
demographic_FA

Subject ID  Age  Sex  Group
0      CON3140   37    2      1
1      CON3891   33    2      1
2      CON4664   40    2      1
3      CON7009   21    1      1
4      CON7024   59    1      1
..         ...  ...  ...    ...
110    RPG9102   42    2      2
111    RPG9103   37    1      2
112    RPG9119   41    1      2
113    RPG9121   51    1      2
114    RPG9126   56    1      2

[115 rows x 4 columns]

In [50]:
print(demographic_CT['Subject ID'])

0       CON9225
1       CON9229
2       CON9231
3      GASP3037
4      GASP3040
         ...   
103     RPG9019
104     RPG9102
105     RPG9119
106     RPG9121
107     RPG9126
Name: Subject ID, Length: 108, dtype: object


In [51]:
print(demographic_FA['Subject ID'])

0      CON3140
1      CON3891
2      CON4664
3      CON7009
4      CON7024
        ...   
110    RPG9102
111    RPG9103
112    RPG9119
113    RPG9121
114    RPG9126
Name: Subject ID, Length: 115, dtype: object


For the **DWI** data there is a total of N=123 participants.

<font color='red'>**NOTE:**</font> **In my case, the excel file format had to be in "xls" for it to be converted into a pandas data frame. The formats "xlsm" and "xlsx" did not work for me. In case you get an import error saying "Missing optional dependency "xlrd"", run the following code below in bash. For further help, click [here](https://datatofish.com/read_excel/).**

In [52]:
#use ! to run command in bash; remove the hashtag below to run the code

#!pip install xlrd

Now, having demographic data subsets of both **CT** and **DWI** data, they can be explored. The documentation on [figshare](https://figshare.com/articles/dataset/Data_for_Functional_MRI_connectivity_accurately_distinguishes_cases_with_psychotic_disorders_from_healthy_controls_based_on_cortical_features_associated_with_brain_network_development_/12361550) provides label information for Sex (1=male, 2=female) and Group (1=control, 2=case).

In [53]:
#n of control and patients
print(demographic_CT['Group'].value_counts()) 

1    80
2    28
Name: Group, dtype: int64


In [54]:
#n of males and females
print(demographic_CT['Sex'].value_counts()) 

1    60
2    48
Name: Sex, dtype: int64


In [55]:
#get age information
demographic_CT['Age'].describe()

count    108.000000
mean      31.231481
std       10.911373
min       18.000000
25%       22.000000
50%       29.000000
75%       38.250000
max       64.000000
Name: Age, dtype: float64

To sum up the demographic data, for the **CT** data within the Dublin sample there is n=80 controls and n=28 patients, a total of N=108 participants of which n=60 are males and n=48 females. The mean age is M=31.23 years with a standard deviation of SD=10.91 years. The age ranges from min=18 years to max=64 years.

In [56]:
#n of control and patients
print(demographic_DWI['Group'].value_counts()) 

NameError: name 'demographic_DWI' is not defined

In [ ]:
#n of males and females
print(demographic_DWI['Sex'].value_counts()) 

In [ ]:
#get age information
demographic_DWI['Age'].describe()

The **DWI data** subset for the Dublin sample consists of a total of N=123 particpants with n=56 being control and n=64 patients. There are n=66 males and n=57 females. The mean age is M=28.46 years with a standard deviation of SD=8.52 years. The age ranges from min=17 years to max=50 years.

## 2. Exploring the different modalities

In the following chapter, the different modalities are explored. Before getting a deeper look at the data, the derived form of the Desikan-Killiany Atlas with 308 cortical regions that is used for the **macro- and micro-structural data** is visualized.
As already mentioned above, there is a [github repository](https://github.com/RafaelRomeroGarcia/subParcellation) for that specific atlas with the required files therein. Since there is a poor documentation of the files and their content, it might be confusing which of them is actually required for visualization purposes. 


If you click on the repository link, the first folder with the title "500mm parcellation (308 regions)" is the relevant one. In the folder itself, there are two text files (.txt) with the coordinates and names of the 308 regions and two NIFTI files (.nii.gz). These are the files, that I used for visualization. For that, I used the nibabel and nilearn modules. 

In [58]:
#import relevant modules

import nibabel as nb
from nilearn import plotting

In [59]:
#load NIFTI file

atlas_path = os.path.join(os.pardir,'data','500.aparc_cortical_consecutive.nii')

atlas = nb.load(atlas_path)

FileNotFoundError: No such file or no access: '../data/500.aparc_cortical_consecutive.nii'

In [ ]:
#plot atlas
plotting.plot_roi(atlas, draw_cross = False, annotate = False, colorbar=True, cmap='Paired', title="Cortical parcellation with 308 cortical regions")

There is also a.txt file that contains all the 308 cortical region names. In the following, the .txt file is read.

In [ ]:
#read the .txt files with region names

with open('../data/atlas/308_regions_names.txt') as f:
    lines = f.readlines()

In [ ]:
#region names
lines

In [ ]:
#proving 308 regions 
len(lines)

As we can see above, the 308 cortical region names are displayed in a chronological order from 1 to 308 for both left and right hemisphere. Each of the regions mostly consists of several parts as the names indicate.
Now, having visualized the used atlas with its regions and according labels, we can explore our first modality. 

### 2.1 Macro-structural data: cortical thickness (CT)

For the **macro-structural data**, T1-weighted MPRAGE images were used. The images were aquired with MPRAGE sequences, capturing high tissue contrast and providing high spatial resolution with whole brain coverage in a short scan time [(Wang et al.,2014)](https://doi.org/10.1371/journal.pone.0096899). These were already preprocessed by a prior pipeline [(Withtaker et al., 2016)](https://pubmed.ncbi.nlm.nih.gov/27457931/), using the command from FreeSurfer called "recon-all" [(Dale et al., 1999)](https://pubmed.ncbi.nlm.nih.gov/9931268/). The surface was parcellated according to the above displayed atlas into 308 regions. For each of those regions cortical thickness was estimated.

**Cortical thickness (CT)** is a measurement of the *width of gray matter* of the cerebral cortex, whereas *gray matter* marks the area from the pial surface to the internal surface (where *white matter* begins). It has been shown that **CT** correlates with the diagnosis and prognosis of several psychiatric and neurological conditions [(Tahedl, 2020)](https://translational-medicine.biomedcentral.com/articles/10.1186/s12967-020-02317-9).

In [ ]:
#load estimated CT data

CT_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_thickness_Dublin.csv')
CT_Dublin = pd.read_csv(CT_Dublin_path)


In [ ]:
CT_Dublin

In [ ]:
#adjust data frame

CT_Dublin_ad = CT_Dublin.drop(['Subject ID','Age', 'Sex'], axis=1)
CT_Dublin_ad

In [ ]:
#compute mean for patients and control in each brain region

df_mean = CT_Dublin_ad.groupby('Group').mean()

In [ ]:
df_mean

In [ ]:
#mean of cortical thickness for control and patients

df_mean.mean(axis=1)

As the mean values indicate, the control group has higher mean in **CT** over all brain areas compared to the patients group. Although the difference does not seem too big at first glance, visualization might help to have a better imagination of the data. Before visualizing the difference in mean values for control and patients, it might be helpful to reshape the dataframe as the columns indicate the respective groups and the rows each brain region.

In [ ]:
#switch colums and rows

df_trans = df_mean.T

In [ ]:
df_trans

In [ ]:
#import modules for visualization

import seaborn as sb
import matplotlib.pyplot as plt

In [ ]:
#plot the CT

plt.figure(figsize=(8,6))
ax = sb.violinplot(data=df_trans)
ax.set(xlabel='Groups', ylabel='Cortical Thickness')
plt.title("Distribution of means of Cortical Thickness for Controls vs. Patients", pad = '20')

The violin plot reveals that the overall cortical thickness in the patient group is lower compared to the controls. 

Group differences between metrics for different brain areas 

To get a broad idea and see, whether there is an effect in certain brain areas in a specific direction compared both groups, the differences in every brain region between the patient and control group is computed and plotted. 

In [ ]:
#compute difference

df_trans['diff'] = df_trans.iloc[:,0] - df_trans.iloc[:,1]

In [ ]:
df_trans

In [ ]:
#plot the difference 

# Initialize layout
fig, ax = plt.subplots(figsize = (20, 20))

# Make histogram
ax.hist(df_trans["diff"],bins= 100, edgecolor = 'black')

plt.title('My title')
plt.xlabel('categories')
plt.ylabel('values')

### 3.2 Micro-structural data: mean diffusivity (MD) and fractional anisotropy (FA)

The **micro-structural** data contains diffusion weighted images (DWI) from which regional cortical measures such as ***mean diffusivity (MD)*** and ***fractional anisotropy (FA)*** were estimated. 

In general, Diffusion MRI measures white matter fibres which makes it feasible to examine connections between different regions. For that, we look at how water diffuses in the brain which again provides information of the brain itself. The diffusion of water can be visualized as cloud if points which again can be approximated with a tensor model. Since there is a distinction in isotropic (characteristics are similar in all directions) vs anisotropic (characteristics e.g. faster in a given direction) diffusion, the tensor model might differ.

**MD** and **FA** are central characteristics of tensors. **MD** indicates how much diffusion there is inside a voxel. **FA** is a measurement of the anisotropy of diffusion with a value range between 0 and 1. While the a **FA** value of 0 stands for isotropic diffusion, the **FA** value of 1 indicates anisotropic diffusion.

In [ ]:
#read MD and FA datasets


MD_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_MD_cortexAv_mean_Dublin.csv')

FA_Dublin_path = os.path.join(os.pardir, 'data', 'PARC_500.aparc_FA_cortexAv_mean_Dublin.csv')


MD_Dublin= pd.read_csv(MD_Dublin_path)
FA_Dublin = pd.read_csv(FA_Dublin_path)

In [ ]:
MD_Dublin

In [ ]:
#adjust data frame
MD_Dublin_ad = MD_Dublin.drop(['Subject ID','Age', 'Sex'], axis=1)

In [ ]:
MD_Dublin_ad

In [ ]:
MD_Dublin_ad.describe()

In [ ]:
#compute mean for patients and control in each brain region

MD_df_mean = MD_Dublin_ad.groupby('Group').mean()

In [ ]:
MD_df_mean

In [ ]:
#mean of mean diffusivity for control and patients

MD_df_mean.mean(axis=1)

The values indicate descriptively that the control group has a lower average mean diffusivity compared to patients.

In [ ]:
#switch colums and rows

MD_df_trans = MD_df_mean.T

In [ ]:
MD_df_trans

In [ ]:
#plot the MD

plt.figure(figsize=(8,6))
ax = sb.violinplot(data=MD_df_trans)
ax.set(xlabel='Groups', ylabel='Mean Diffusivity')
plt.title("Distribution of Mean Diffusivity Means for Controls vs. Patients", pad = '20')

Group differences

In [ ]:
#compute difference

MD_df_trans['diff'] = MD_df_trans.iloc[:,0] - MD_df_trans.iloc[:,1]

In [ ]:
MD_df_trans

In [ ]:
#plot the difference 

# Initialize layout
fig, ax = plt.subplots(figsize = (9, 9))

# Make histogram
ax.hist(MD_df_trans["diff"], edgecolor = 'black')


In [ ]:
FA_Dublin

In [ ]:
#adjust data frame

FA_Dublin_ad = FA_Dublin.drop(['Subject ID','Age', 'Sex'], axis=1)

In [ ]:
FA_Dublin_ad

In [ ]:
FA_Dublin_ad.describe()

In [ ]:
#compute mean for patients and control in each brain region

FA_df_mean = FA_Dublin_ad.groupby('Group').mean()

In [ ]:
FA_df_mean

In [ ]:
#mean of fractional anisotropy for control and patients

FA_df_mean.mean(axis=1)

The values seem to indicate an isotropic diffusion in both control and patient groups.

In [ ]:
#switch colums and rows

FA_df_trans = FA_df_mean.T

In [ ]:
FA_df_trans

In [ ]:
#plot the FA

plt.figure(figsize=(8,6))
ax = sb.violinplot(data=FA_df_trans)
ax.set(xlabel='Groups', ylabel='Fractional Anisotropy')
plt.title("Distribution of Fractional Anisotropy Means for Controls vs. Patients", pad = '20')

In [ ]:
#compute difference

FA_df_trans['diff'] = FA_df_trans.iloc[:,0] - FA_df_trans.iloc[:,1]

In [ ]:
FA_df_trans

In [ ]:
#plot the difference 

# Initialize layout
fig, ax = plt.subplots(figsize = (9, 9))

# Make histogram
ax.hist(FA_df_trans["diff"], edgecolor = 'black')

## 3.1 DTI Networks

First of all, I am going to explore the DTI network data, followed by the regional MD and FA values and lastly the CT data.

The data that contains the DTI networks are available as matlab files. In the following, it is depicted how the matlab files can be downloaded and read.

####  3.1.1 Download the .mat files

If you click on the name of the matlab datafile on figshare , it then only shows you the preview of the matlab file and the link of that is for the respective preview. If you want to copy the link of the matlab file itself, you have to right-click on the datafile name and then copy the link.

In [ ]:
import urllib.request

In [ ]:
print('Beginning file download with urllib2...')

url = "https://figshare.com/ndownloader/files/22782440"
urllib.request.urlretrieve(url, r'../data/DTI_Dublin.mat')

#### 3.1.2 Read the .mat files

In [ ]:
import scipy.io

**Dublin**

In [ ]:
DTI_Dublin = scipy.io.loadmat(r'../data/DTI_Dublin.mat')

In [ ]:
DTI_Dublin.keys()

In [ ]:
DTI_Dublin['nostreamlines_new'].shape

123 Probanden mit DTI Matrizen.
Wie sehen DTI Matrizen aus?

In [ ]:
DTI_Dublin['nostreamlines_new'][0].shape

In [ ]:
DTI_Dublin['nostreamlines_new'][0][0].shape

Zwischen 308 Regionen basierend auf DTI Daten Werte (siehe course website, nochmal durchlesen!!)

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.figure(figsize=(10,7))
sns.heatmap(DTI_Dublin['nostreamlines_new'][0][0], xticklabels=False, cmap='rocket')